feat: ciphertext generation algorithm

Jatewo opened the pull request:
This pull request introduces a homophonic substitution cipher implementation, including the core cipher logic, homophone assignment based on English letter frequencies, and a comprehensive suite of tests. It also enhances the Python project configuration for testing.

### Core Cipher Functionality

* Added the `Cipher` class in `cipher.py`, which validates input, generates a homophonic key based on letter frequencies and difficulty, enciphers plaintext, and provides JSON and string representations.
* Implemented the `extract_homophones`, `adjust_homophones`, and `add_noise` functions in `homophones.py` to distribute cipher symbols among letters according to their frequency, with randomness and validation.

### Testing

* Added extensive tests for the cipher and homophone logic in `test_cipher.py` and `test_homophones.py`, covering input validation, key generation, ciphertext correctness, and edge cases. [[1]](diffhunk://#diff-73ddfa1579b46b001abf7ffe4f9e34da9a505f9b82c8b6d497aca5bec5aa83a3R1-R83) [[2]](diffhunk://#diff-65e0f3ea762e26871a0b92db0c303d963cb16923060902c556d3b10dc7bb5f46R1-R89)

### Project Configuration

* Updated `pyproject.toml` to include `pytest` and `pytest-cov` as dependencies, set test paths, and enabled coverage reporting for easier test execution and quality assurance.

Jatewo committed:
feat: Add homophonic substitution cipher homophones implementation

Introduces a Cipher class for generating homophonic substitution cipher keys based on English letter frequencies and a randomized difficulty level. Includes utility functions for adjusting homophone counts, adding noise, and generating difficulty, along with letter frequency data.

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

Jatewo committed:
Merge branch 'dev' into enciphering-algorithm

Jatewo committed:
feat: Add homophone extraction utility for encipherment

Introduces functions to calculate and adjust homophone counts for cipher symbols based on English letter frequencies, including random noise for variability. This utility will support homophonic substitution cipher implementations.

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

Jatewo committed:
refactor: Refactor Cipher class and delegate homophone extraction

Refactored the Cipher class to delegate homophone extraction to the new extract_homophones function from the homophones module. Removed internal noise and adjustment logic, simplified key generation, and improved encapsulation. Added __json__ and __str__ methods for better object representation. Updated main block for demonstration.

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

Jatewo committed:
depend: uv.lock update

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

Jatewo committed:
refactor: Move encipherment modules to src directory

Renamed cipher.py and homophones.py from encipherment/ to src/encipherment/. Updated import in cipher.py to use relative import for extract_homophones.

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

Jatewo committed:
dep: Add pytest and pytest-cov to dependencies and config

Added 'pytest' and 'pytest-cov' to the dependencies in pyproject.toml. Updated pytest configuration to specify test paths and coverage options for improved testing and reporting.

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

Jatewo committed:
test: Add tests for Cipher class input validation

Introduces tests to verify that the Cipher class correctly handles legal and illegal plaintext inputs. Legal input is a long lowercase string, while illegal inputs include uppercase letters, numbers, spaces, dashes, and punctuation. The tests check for correct attribute initialization and proper error raising.

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

Jatewo committed:
clean: Remove main block and test code from cipher.py

Deleted the __main__ block and its test instantiation of the Cipher class, leaving only the class definition. This streamlines the module for import and use elsewhere.

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

Jatewo committed:
test: Expand Cipher class tests with additional cases

Added tests for key homophones count, ciphertext number range, ciphertext length, JSON serialization, and string representation of the Cipher class. Improved formatting and consistency in existing tests.

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

Jatewo committed:
fix: Refactor homophones module for clarity and validation

Improved docstrings and argument validation in extract_homophones, adjust_homophones, and add_noise functions. Added a check to ensure cipher_symbols is at least 26, and reformatted code for better readability and maintainability.

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

Jatewo committed:
test: Add input validation and tests for homophones module

Added a ValueError check in adjust_homophones to ensure cipher_symbols is at least 26. Introduced comprehensive tests for extract_homophones, add_noise, and adjust_homophones, including validation of error handling for invalid input values.

Co-Authored-By: MaizeHH <92512497+MaizeHH@users.noreply.github.com>

TheMOHD3886 reviewed (approved):
lgtm 🤙

Jatewo committed:
CI/CD: Add GitHub Actions for linting and testing

Introduces two workflow files: one for running Ruff linter and another for executing PyTest on pushes and pull requests to main and dev branches. Both workflows use Python 3.12 and the uv package manager for dependency management.

Jatewo committed:
clean: formatting

Jatewo committed:
fix: Add ruff as a dev dependency and clean up imports

Added ruff (>=0.13.1) to the dev dependency group in pyproject.toml and updated the lock file accordingly. Also removed an unused import of Decimal from cipher.py.

Mafusn commented on pyproject.toml:
```diff
@@ -4,7 +4,14 @@ version = "0.1.0"
 readme = "README.md"
 requires-python = ">=3.12"
-dependencies = []
+dependencies = ["pytest", "pytest-cov"]
```
These should be moved to dev-dependencies together with ruff "uv add --dev pytest pytest-cov ruff"

Mafusn commented on test/encipherment/test_homophones.py:
```diff
@@ -0,0 +1,89 @@
+
+
+@pytest.fixture
+def homphone_dict():
```
I agree 💅

Mafusn reviewed (approved):
Looks good, but maybe move the packages to dev-dependencies

Jatewo committed:
fix: Move pytest dependencies to dev group and fix test typo

Removed pytest and pytest-cov from main dependencies in pyproject.toml and added them to the dev dependency group with version constraints. Fixed a typo in test_homophones.py by renaming 'homphone_dict' to 'homophone_dict' and updated all references accordingly.
